File: week3/8bitbully/minEatingSpeed.cpp
// 875. Koko Eating Bananas

#include <cstdint>
#include <vector>

using namespace std;

int minEatingSpeed(vector<int> &piles, int h) {
  int64_t left = 1;
  int64_t right = 1e9;

  auto find = [&](int64_t &mid) {
    int64_t hours = 0;
    for (auto &p : piles) {
      hours += (p + mid - 1) / mid;
    }
    return hours;
  };

  while (left <= right) {
    int64_t mid = left + (right - left) / 2;

    if (find(mid) <= h) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
}

File: week3/MarshalLee/MinEatingSpeed.cs
using System;

public static int MinEatingSpeed(int[] piles, int h)
{
    int left = 1;
    int right = piles.Max();

    while (left < right)
    {
        int mid = left + (right - left) / 2;
        int totalHours = CalculateTotalHours(piles, mid);

        if (totalHours > h)
        {
            left = mid + 1;
        }
        else
        {
            right = mid;
        }
    }

    return left;
    static int CalculateTotalHours(int[] piles, int speed)
    {
        int totalHours = 0;

        foreach (int pile in piles)
        {
            totalHours += (int)Math.Ceiling((double)pile / speed);
        }

        return totalHours;
    }
    //int[] possibleKs = Enumerable.Range(1, piles.Max()).ToArray();
    //int res = h;

    //int left = 0, right = possibleKs.Length - 1;
    //while (left <= right)
    //{
    //    int index = (left + right) / 2;
    //    int currK = possibleKs[index];

    //    int totalHours = 0;
    //    foreach (int pile in piles)
    //    {
    //        totalHours += (int)Math.Ceiling((double)pile / currK);
    //    }

    //    if (totalHours <= h)
    //    {
    //        res = Math.Min(res, currK);
    //        right = index + 1;
    //    }
    //    else if(totalHours >= h)
    //    {
    //        left = index - 1;
    //    }
    //}
    //return res;
}